rfcs/0097-unify-packed-and-object.md

```diff
@@ -0,0 +1,677 @@
+- D3b: 8 byte per Object, 8 bytes ptr + 24 bytes(boxed object) if original value is not boxed.
+- D3c: 8 bytes per Object, 32 bytes per POD(small-str)
+- D3c-variant: 8 bytes per Object, 8 bytes per small integer and char that have static pool.
+
+Accessing efficiency:
+
+D3b may have overhead for small-str and POD. D3a and D3c are small-str friendly.
+
+### D4: PackedFunc Convention
+
+In this section, we revisit the PackedFunc convention under the new context. The update to the C++ PackedFunc API will run as follows:
+
+- TVMArgs now stores ptr of AnyView, TVMArgs[i] will now return AnyView
+- Alias: TVMRetValue = AnyValue, TVMArgValue = AnyView
+
+These changes are invisible to the users as long as they use the same source library. We will immiedately gain the ability to do universal switching when defining PackedFunc.
+
+- A developer can choose to develop solely on ObjectRef, in this case automatic conversion happens when we turn AnyView and request Object. We expect most of the compiler development to be in this mode.
+- A developer can choose to develop runtime functions that contains AnyValue and AnyView to take benefit of stack values. This will have benefit that intermediate values can store small-str. We anticipate some of the builtin runtime to operate on this mode.
+- String will be backed by both small-str and object for efficiency, which we expect to help compiler as well(as there are a lot of small names).
+- Both runtime and compiler will be backed by universal container object, which allows us to simplify the automatic conversion in FFI (you take a python tuple and would expect an Array).
+
+There are several design choices in terms of C API convention in light of this new proposal. They will affect the internal data layout of TVMArgs.
+
+- **D4a:** Current C API:
+    
+    ```cpp
+    int TVMCPackedFunc(PackedFuncHandle handle,
+                       int num_args, int* type_codes, TVMValue* values,
+                       int* out_tcode, TVMValue* out_value);
+    ```
+    
+    - This would require packing unpacking the typecode. Note that small string passing wont work because of the lack of the seq_len padding field.
+- **D4b:** Combine type code and TVMValue
+    
+    ```cpp
+    
+    // TVMAnyView and TVMAnyValue follows the same layout as AnyPodBaseValue_
+    int TVMCPackedFunc(PackedFuncHandle handle,
+                       int num_args, TVMAnyView* args, 
+                       TVMAnyValue* out);
+    ```
+    
+    This approach combines code and value into a single entity, this would mean a change of ABI convention in generated code. This approach makes it possible to directly return a small-str without boxing it (however if it is faced in python frontend, likely we still need to box it to str).
+    
+    We will introduce adapters to support the old calling convention, which constructs TVMAnyView on the stack then pass things over in the transitioning period.
```

Comment:
What's the planned naming conventions for this and how do they eventually get transitioned to just being `TVMCPackedFunc` again?

Reply:
The naming convention is being as proposed TVMAnyValue. The particular signature is a C signature for the frontend API.

Reply:
Added an expanded explaination